src/realm/object-store/sync/app.cpp

```diff
@@ -774,34 +811,117 @@ void App::post(std::string&& route, util::UniqueFunction<void(util::Optional<App
+            // if this is the first time we have received app metadata, the
+            // original request will not have the correct URL hostname for
+            // non global deployments.
+            auto app_metadata = anchor->m_sync_manager->app_metadata();
+            if (app_metadata && app_metadata->deployment_model != "GLOBAL" &&
+                request.url.rfind(anchor->m_base_url, 0) != std::string::npos) {
+                request.url.replace(0, anchor->m_base_url.size(), app_metadata->hostname);
+            }
+
+            // Retry the original request with the updated url
+            anchor->m_config.transport->send_request_to_server(
+                std::move(request), [completion = std::move(completion), anchor = std::move(anchor)](
+                                        const Request& request, const Response& error) mutable {
+                    if (error.http_status_code == 301) {
+                        anchor->handle_redirect_response(request, error, std::move(completion));
+                    }
+                    else {
+                        completion(std::move(error));
+                    }
+                });
+        },
+        new_hostname);
+}
+
+void App::handle_redirect_response(const Request& request, const Response& error,
+                                   util::UniqueFunction<void(const Response&)>&& completion)
+{
+    // Permanent redirect - get the location and init the metadata again
+    auto location = error.headers.end();
+    auto needle = std::string("location");
+    // Look for case insensitive redirect "location" in headers
+    for (auto it = error.headers.begin(); it != error.headers.end(); ++it) {
+        if (std::equal(it->first.begin(), it->first.end(), needle.begin(), needle.end(), [](char a, char b) {
+                return tolower(a) == tolower(b);
+            })) {
+            location = it;
+            break; // value found, stop looping
         }
+    }
 
-        anchor->m_config.transport->send_request_to_server(std::move(request), std::move(completion));
-    });
+    if (location != error.headers.end()) {
+        constexpr auto https_len = std::char_traits<char>::length("https://");
+        auto new_url = location->second;
+        if (new_url.length() > 0) {
+            // Trim off any trailing path/anchor/query string
+            size_t split = new_url.find_first_of("/#?", https_len);
+            if (split != std::string::npos) {
+                new_url.erase(split);
+            }
+
+            auto request_c = std::move(request);
+
+            if (request_c.max_redirects) {
+                // Make sure we don't do too many redirects (20 is an arbitrary number)
+                if (*(request_c.max_redirects) > 20) {
+                    auto error_c = std::move(error);
+                    error_c.http_status_code = 399; // custom status code for too many redirects
```

Comment:
hmm. can we set the custom_error_code to something appropriate here rather than making up a non-standard http status.